cmd/kubeadm/app/phases/controlplane/volumes.go

```diff
@@ -86,37 +86,52 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
 		mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, caCertsPkiVolumeName, caCertsPkiVolumePath, caCertsPkiVolumePath, true, &hostPathDirectoryOrCreate)
 	}
 
+	if len(cfg.APIServerExtraVolumes) != 0 {
+		mounts.NewHostPathMount(kubeadmconstants.KubeAPIServer, cfg.APIServerExtraVolumes.Name, cfg.APIServerExtraVolumes.HostPath, cfg.APIServerExtraVolumes.MountPath, true, &hostPathDirectoryOrCreate)
+	}
+
+	if len(cfg.ControllerManagerExtraVolumes) != 0 {
+		mounts.NewHostPathMount(kubeadmconstants.KubeControllerManager, cfg.ControllerManagerExtraVolumes.Name, cfg.ControllerManagerExtraVolumes.HostPath, cfg.ControllerManagerExtraVolumes.MountPath, true, &hostPathDirectoryOrCreate)
+	}
+
+	if len(cfg.SchedulerExtraVolumes) != 0 {
+		mounts.NewHostPathMount(kubeadmconstants.KubeScheduler, cfg.SchedulerExtraVolumes.Name, cfg.SchedulerExtraVolumes.HostPath, cfg.SchedulerExtraVolumes.MountPath, true, &hostPathDirectoryOrCreate)
+	}
 	return mounts
 }
 
 // controlPlaneHostPathMounts is a helper struct for handling all the control plane's hostPath mounts in an easy way
 type controlPlaneHostPathMounts struct {
-	volumes      map[string][]v1.Volume
-	volumeMounts map[string][]v1.VolumeMount
+	volumes      map[string]map[string]v1.Volume
```

Comment:
is there any simple way we can make this more readable? At least document what strings should be where?

Reply:
Agreed. Really ugly. I can fix this up

Reply:
not sure I see the comment improvement ;)